File: day_2/part_1.py
def is_invalid_id(n: int) -> bool:
    s = str(n)
    length = len(s)

    if length % 2 != 0:
        return False

    if s[0] == "0":
        return False

    mid = length // 2
    return s[:mid] == s[mid:]


def solve(input: str) -> int:
    ranges = []
    for range_str in input.strip().split(","):
        range_str = range_str.strip()
        if range_str:
            start, end = map(int, range_str.split("-"))
            ranges.append((start, end))

    sum = 0

    for start, end in ranges:
        for i in range(start, end + 1):
            if is_invalid_id(i):
                sum += i

    return sum


if __name__ == "__main__":
    with open("day_2.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: go/day_5/part_1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(input string) int {
	sections := strings.Split(strings.TrimSpace(input), "\n\n")

	ranges := [][2]int{}
	for _, line := range strings.Split(strings.TrimSpace(sections[0]), "\n") {
		if strings.TrimSpace(line) != "" {
			parts := strings.Split(line, "-")
			start, _ := strconv.Atoi(parts[0])
			end, _ := strconv.Atoi(parts[1])
			ranges = append(ranges, [2]int{start, end})
		}
	}

	ingredient_ids := []int{}
	for _, line := range strings.Split(strings.TrimSpace(sections[1]), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			id, err := strconv.Atoi(line)
			if err == nil {
				ingredient_ids = append(ingredient_ids, id)
			}
		}
	}

	fresh_count := 0
	for _, id := range ingredient_ids {
		is_fresh := false
		for _, r := range ranges {
			start, end := r[0], r[1]
			if start <= id && id <= end {
				is_fresh = true
				break
			}
		}
		if is_fresh {
			fresh_count++
		}
	}

	return fresh_count
}

func main() {
	input, err := os.ReadFile("../inputs/day_5.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
